Add doc for DeletionPolicy and Finalizers (#14719)
kubernetes/website: content/en/docs/concepts/storage/volume-snapshots.md

VolumeSnapshots will remain unbound indefinitely if a matching VolumeSnapshotContent does not exist. VolumeSnapshots will be bound as matching VolumeSnapshotContents become available.

### Snapshot Object in Use Protection

> The purpose of the Snapshot Object in Use Protection feature is to ensure that in-use snapshot API objects are not removed from the system (as this may result in data loss). There are two cases that require “in-use” protection:

Reviewer: ```suggestion
The _Snapshot Object in Use Protection_ feature lets you ensure that in-use snapshot API objects are not removed from the system (as this may result in data loss). There are two cases that require “in-use” protection:
```

Author: Done

> * If a volume snapshot is in active use by a persistent volume claim as a source to create a volume.

Reviewer: ```suggestion
* If a volume snapshot is in active use by a {{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}} (PVC) as a source to create a Volume.
```

Author: Done

* If a `VolumeSnapshotContent` API object is bound to a `VolumeSnapshot` API object, the content object is considered in use.

> If a user deletes a `VolumeSnapshot` API object in active use by a PVC, the `VolumeSnapshot` object is not removed immediately. Instead, removal of the `VolumeSnapshot` object is postponed until the `VolumeSnapshot` is no longer actively used by any PVCs. Similarly, if an admin deletes a `VolumeSnapshotContent` that is bound to a `VolumeSnapshot`, the `VolumeSnapshotContent` is not removed immediately. Instead, the `VolumeSnapshotContent` removal is postponed until the `VolumeSnapshotContent` is not bound to the `VolumeSnapshot` object.

Reviewer: ```suggestion
If you delete a `VolumeSnapshot` API object that a PVC is actively using, the `VolumeSnapshot` object is not removed immediately. Instead, removal of the `VolumeSnapshot` object is postponed until the `VolumeSnapshot` is no longer actively used by any PVCs. Similarly, if an admin deletes a `VolumeSnapshotContent` that is bound to a `VolumeSnapshot`, the `VolumeSnapshotContent` is not removed immediately. Instead, the `VolumeSnapshotContent` removal is postponed until the `VolumeSnapshotContent` is not bound to the `VolumeSnapshot` object.
```

Author: Done

### VolumeSnapshotContent Deletion/Retain Policy

> The Kubernetes snapshot APIs are similar to the PV/PVC APIs: just like a volume is represented by a bound PVC and PV pair, a snapshot is represented by a bound `VolumeSnapshot` and `VolumeSnapshotContent` pair.

Reviewer: ```suggestion
The Kubernetes snapshot APIs are similar to the PV/PVC APIs: just like a volume is represented by a bound PVC and PV pair, a snapshot is represented by a bound VolumeSnapshot and VolumeSnapshotContent pair.
```

Author: Done

With PV/PVC pairs, when a user is done with a volume, they can delete the PVC. And the reclaim policy on the PV determines what happens to the PV (whether it is also deleted or retained).

> In the initial alpha release, snapshots did not support the ability to specify a reclaim policy. Instead when a snapshot object was deleted it always resulted in the snapshot being deleted. In Kubernetes v1.13, a snapshot content `DeletionPolicy` was added. It enables an admin to configure what happens to a `VolumeSnapshotContent` after the `VolumeSnapshot` object it is bound to is deleted. The `DeletionPolicy` of a volume snapshot can either be `Retain` or `Delete`. If the value is not specified, the default depends on whether the `VolumeSnapshotContent` object was created via static binding or dynamic provisioning.

Reviewer 2: Break the lines at a reasonable column would make the review a bit easier.

"snapshots did not support the ability to specify a reclaim policy." This sounds a wordy sentence though I'm not a native speaker. You could simply say "snapshots did not support reclaim policies".

Reviewer 2: "Instead when a snapshot object was deleted it always resulted in the snapshot being deleted."

=>

"When a snapshot resource is deleted, the snapshot is deleted as well".

Reviewer 3: I agree with your suggestion, with just a slight update to the tense, so from:

"In the initial alpha release, snapshots did not support the ability to specify a reclaim policy. Instead when a snapshot object was deleted it always resulted in the snapshot being deleted."

to:

"In the initial alpha release, snapshots did not support reclaim policies. When a snapshot resource was deleted, the snapshot was deleted as well."

Author: Done

#### Retain

The `Retain` policy allows for manual reclamation of the resource. If a `VolumeSnapshotContent` is statically created and bound, the default `DeletionPolicy` is `Retain`. When the `VolumeSnapshot` is deleted, the `VolumeSnapshotContent` continues to exist and the `VolumeSnapshotContent` is considered “released”. But it is not available for binding to other `VolumeSnapshot` objects because it contains data. It is up to an administrator to decide how to handle the remaining API object and resource cleanup.

#### Delete

A `Delete` policy enables automatic deletion of the bound `VolumeSnapshotContent` object from Kubernetes and the associated storage asset in the external infrastructure (such as an AWS EBS snapshot or GCE PD snapshot, etc.). Snapshots that are dynamically provisioned inherit the deletion policy of their `VolumeSnapshotClass`, which defaults to `Delete`. The administrator should configure the `VolumeSnapshotClass` with the desired retention policy. The policy may be changed for individual `VolumeSnapshotContent` after it is created by patching the object.

> The following example demonstrates how to check the deletion policy of a dynamically provisioned `VolumeSnapshotContent`.

Reviewer: Would this work better as a Task page?

Author: I'm not sure.  Does the Task page have a link to the Concept page?

Reviewer 4: @xing-yang When you create a new task page, you can certainly link to the relevant concept page(s).

```

> $ kubectl create -f ./examples/kubernetes/demo-defaultsnapshotclass.yaml

Reviewer 2: where is the yaml file?

Reviewer: To follow the style guide, [don't include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt).

Author: @tengqm Good catch.  I need to add the example yaml file here.  Should I create a new folder called "storage" under website/content/en/examples and put the yaml file there?

Author: Done
